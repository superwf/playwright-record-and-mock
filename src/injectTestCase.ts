import fs from 'fs'

import { parse } from '@babel/parser'
import traverse from '@babel/traverse'
import generate from '@babel/generator'
import template from '@babel/template'
import type { Statement } from '@babel/types'
import { isIdentifier, isCallExpression, isBlockStatement } from '@babel/types'

import { getTestCaseFilePath } from './tool'
import { InjectResult } from './type'

export const injectTestCase = ({ caseName, outDir }: { caseName: string; outDir: string }): InjectResult => {
  const testCaseFile = getTestCaseFilePath(outDir, caseName)
  const source = fs.readFileSync(testCaseFile, 'utf8')

  const ast = parse(source, {
    sourceType: 'module',
  })

  traverse(ast, {
    ImportDeclaration(path) {
      if (path.node.source.value === '@playwright/test') {
        const t = template.ast(`
      import { mock } from 'playwright-record-and-mock'
      `)
        path.insertAfter(t)
        path.skip()
      }
    },
    ArrowFunctionExpression(path) {
      if (isCallExpression(path.parentPath.node)) {
        const { callee } = path.parentPath.node
        // 判断是否为 test 的调用参数函数
        if (isIdentifier(callee) && callee.name === 'test') {
          if (isBlockStatement(path.node.body)) {
            const { body } = path.node.body
            if (Array.isArray(body)) {
              const alreadyMock = body.some(node =>
                source.slice(node.start as number, node.end as number).includes('mock(page)'),
              )
              if (!alreadyMock) {
                body.unshift(template.ast(`await mock(page, '${caseName}');`) as Statement)
              }
            }
          }
        }
      }
    },
  })
  const injectedCode = generate(ast).code
  return {
    testCaseFile,
    injectedCode,
  }
}
