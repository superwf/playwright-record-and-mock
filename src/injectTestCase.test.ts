import fs from 'fs'
import { injectTestCase } from './injectTestCase'

it('injectTestCase', () => {
  const res = injectTestCase({
    outDir: 'e2e',
    caseName: 'npm',
  })
  expect(res.injectedCode).toMatchSnapshot()
  // const code = fs.readFileSync(testCaseFile, 'utf8')
  // console.log(code)
})
