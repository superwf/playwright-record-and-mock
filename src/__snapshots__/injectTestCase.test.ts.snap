// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`injectTestCase 1`] = `
"import { test, expect } from '@playwright/test';
import { mock } from 'playwright-record-and-mock';
import { mock } from 'playwright-record-and-mock';
test('test', async ({
  page
}) => {
  await mock(page, 'npm');
  await mock(page, 'npm'); // Go to https://www.npmjs.com/

  await page.goto('https://www.npmjs.com/'); // Click #search div >> nth=1

  await page.locator('#search div').nth(1).click(); // Click [placeholder=\\"Search packages\\"]

  await page.locator('[placeholder=\\"Search packages\\"]').click(); // Fill [placeholder=\\"Search packages\\"]

  await page.locator('[placeholder=\\"Search packages\\"]').fill('playwright'); // Press Enter

  await page.locator('[placeholder=\\"Search packages\\"]').press('Enter'); // Click text=playwrightexact match >> h3

  await page.locator('text=playwrightexact match >> h3').click();
  await expect(page).toHaveURL('https://www.npmjs.com/package/playwright'); // Click text=playwright1.24.2 • Public • Published 7 days ago >> span >> nth=0

  await page.locator('text=playwright1.24.2 • Public • Published 7 days ago >> span').first().click();
  await expect(page).toHaveURL('https://www.npmjs.com/package/playwright');
});"
`;
